### PR TITLE
Som devel 20

### PR DESCRIPTION
**Pull request by sorgom**

### CODE REVIEW

File: application/components/COM/TCP_Cons.h [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+private:
+    INT32 mSocket = -1;
+    ComTele mTele;
+};
+
+//  field tcp communication client
+class TCP_Con_Fld : public TCP_Con_Base
+{
+public:
+    inline TCP_Con_Fld() = default;
```
Comment (sorgom) [inline application/components/COM/TCP_Cons.h]: if you provide an instance, then why is the constructor public?

File: application/components/COM/TCP_Cons.h [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+protected:
+    void onAccept() const override;
+    void forward(const ComTele& tele) const override;
+};
+
+//  GUI tcp communication client
+class TCP_Con_Gui : public TCP_Con_Base
+{
+public:
+    inline TCP_Con_Gui() = default;
```
Comment (sorgom) [inline application/components/COM/TCP_Cons.h]: if you provide an instance, then why is the constructor public?

File: application/components/COM/TCP_Cons.h [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+protected:
+    void onAccept() const override;
+    void forward(const ComTele& tele) const override;
+};
+
+//  control tcp communication client
+class TCP_Con_Ctrl : public TCP_Con_Base
+{
+public:
+    inline TCP_Con_Ctrl() = default;
```
Comment (sorgom) [inline application/components/COM/TCP_Cons.h]: if you provide an instance, then why is the constructor public?

File: application/components/COM/TCP_Listeners.h [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+    Tcp_Listener_Base() = default;
+    virtual I_TCP_Con& getCon() const = 0;
+private:
+    INT32 mSocket = -1;
+};
+
+class TCP_Listener_Fld : public Tcp_Listener_Base
+{
+public:
+    inline TCP_Listener_Fld() = default;
```
Comment (sorgom) [inline application/components/COM/TCP_Listeners.h]: if you provide an instance, then why is the constructor public?

File: application/components/COM/TCP_Listeners.h [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+    INSTANCE_DEC(TCP_Listener_Fld)
+    NOCOPY(TCP_Listener_Fld)
+protected:
+    I_TCP_Con& getCon() const override;
+};
+
+class TCP_Listener_Gui : public Tcp_Listener_Base
+{
+public:
+    inline TCP_Listener_Gui() = default;
```
Comment (sorgom) [inline application/components/COM/TCP_Listeners.h]: if you provide an instance, then why is the constructor public?

File: application/components/COM/TCP_Listeners.h [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+    INSTANCE_DEC(TCP_Listener_Gui)
+    NOCOPY(TCP_Listener_Gui)
+protected:
+    I_TCP_Con& getCon() const override;
+};
+
+class TCP_Listener_Ctrl : public Tcp_Listener_Base
+{
+public:
+    inline TCP_Listener_Ctrl() = default;
```
Comment (sorgom) [inline application/components/COM/TCP_Listeners.h]: if you provide an instance, then why is the constructor public?

File: application/components/COM/src/Com.cpp [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+    I_TCP_Listener& listenerGui = IL::getTCP_Listener_Gui();
+    I_TCP_Listener& listenerCtrl = IL::getTCP_Listener_Ctrl();
+
+    I_TCP_Con &clientFld = IL::getTCP_Con_Fld();
+    I_TCP_Con &clientGui = IL::getTCP_Con_Gui();
+    I_TCP_Con &clientCtrl = IL::getTCP_Con_Ctrl();
+
+    const ComSetup& setup = IL::getReader().getComSetup();
+
+    I_TCP& tcp = IL::getTCP();
```
Comment (sorgom) [inline application/components/COM/src/Com.cpp:18]: maybe move up to the other instances

File: application/components/COM/Com.h [hunk 8]
```diff
@@ -15,12 +13,14 @@ class Com : public I_Com
 {
 public:
     inline Com() = default;
```
Comment (sorgom) [inline application/components/COM/Com.h]: if you provide an instance, then why is the constructor public?

File: application/components/COM/src/TCP_Cons.cpp [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+{
+    if (mSocket >= 0)
+    {
+        IL::getTCP().send(mSocket, reinterpret_cast<const CHAR*>(&tele), sizeof(ComTele));
+    }
+    else
+    { pass(); }
+}
+
+//  field client
```
Comment (sorgom) [inline application/components/COM/src/TCP_Cons.cpp]: should be named connection

File: application/components/COM/src/TCP_Cons.cpp [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+{
+    //  TODO: Dispatcher must recall all states
+}
+
+void TCP_Con_Fld::forward(const ComTele& tele) const
+{
+    IL::getDispatcher().fromFld(tele);
+}
+
+//  GUI client
```
Comment (sorgom) [inline application/components/COM/src/TCP_Cons.cpp]: should be named connection

File: application/components/COM/src/TCP_Cons.cpp [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+    //  TODO: Dispatcher resend all states
+    //  call I_Elem::resend() or similar
+}
+
+void TCP_Con_Gui::forward(const ComTele& tele) const
+{
+    IL::getDispatcher().fromGui(tele);
+}
+
+//  control client
```
Comment (sorgom) [inline application/components/COM/src/TCP_Cons.cpp]: should be named connection

File: application/components/SYS/IL.h [hunk 12]
```diff
@@ -6,27 +6,33 @@
...
-#include <ifs/I_Provider.h>
-#include <ifs/I_Reader.h>
-
-#include <BAS/coding.h>
+#include <COM/TCP.h>
+#include <COM/TCP_Cons.h>
+#include <COM/TCP_Listeners.h>
+#include <LCR/LCR_Provider.h>
+#include <SIG/SIG_Provider.h>
+#include <COM/Com.h>
```
Comment (sorgom) [inline application/components/SYS/IL.h]: should be sorted avoid duplicates

File: application/components/SYS/src/Reader.cpp [hunk 13]
```diff
@@ -23,11 +23,12 @@ void Reader::read(const CONST_C_STRING filename) const
     {
         constexpr static stype nCAP = 4;
         constexpr static stype hSize = nCAP * sizeof(UINT32);
+        constexpr static stype minSize = hSize + sizeof(ComSetup);
```
Comment (sorgom) [inline application/components/SYS/src/Reader.cpp:26]: in general there should be a graphical specification of this

File: make/premake5.lua [hunk 14]
```diff
@@ -20,7 +20,7 @@ workspace 'lib_cpputest'
             targetdir 'lib'
             files { files_cpputest_gcc }
             includedirs { includedirs_cpputest }
-            buildoptions { buildoptions_gcc }
+            buildoptions { buildoptions_gcc .. ' -DCPPUTEST_MEM_LEAK_DETECTION_DISABLED' }
```
Comment (sorgom) [inline make/premake5.lua:34]: in general: why is this disabled?

File: make/premake5.lua [hunk 15]
```diff
@@ -97,60 +102,62 @@ workspace 'coverage'
...
-        project 'systests_tests'
+        project 'systemtests_run'
             kind 'ConsoleApp'
             targetdir 'bin'
             libdirs { 'lib' }
-            links { 'systests_app_%{cfg.name}', links_test_gcc }
-            files { files_testenv, files_systest }
-            removefiles { removefiles_systest_test }
+            links { links_test_gcc }
+            files { files_testenv, files_runtests }
```
Comment (sorgom) [inline make/premake5.lua]: files_systemtest ?

File: make/premake5_settings.lua [hunk 16]
```diff
@@ -20,30 +20,33 @@ files_teststeps = {
 
 includedirs_app = {
     '../specification',
-    '../application',
     '../application/components'
 }
 
 includedirs_test = {
     '../testing/testenv',
```
Comment (sorgom) [inline make/premake5_settings.lua]: why is this not  a variable?

File: make/premake5_settings.lua [hunk 17]
```diff
@@ -59,15 +62,20 @@ defines_app = { 'NDEBUG' }
 
 defines_gendata = {
```
Comment (sorgom) [inline make/premake5_settings.lua:67]: this is also relevant for system tests, maybe rename

File: make/premake5_settings.lua [hunk 18]
```diff
@@ -89,4 +97,6 @@ files_cpputest_vs = {
 files_cpputest_gcc = {
     files_cpputest,
     base_cpputest .. '/src/Platforms/Gcc/*.cpp'
-}
\ No newline at end of file
+}
+
+linkoptions_test_gcc = { '-pthread' }
```
Comment (sorgom) [inline make/premake5_settings.lua:105]: why exactly do we need this?

File: make/premake5_vs.lua [hunk 19]
```diff
@@ -19,7 +19,7 @@ include 'premake5_settings.lua'
 --  ============================================================
 buildoptions_vs = '/std:c++17 /MP'
 buildoptions_vs_app = buildoptions_vs .. ' /W4 /wd4100 /wd4103'
-buildoptions_vs_test = buildoptions_vs_app .. ' /wd4127'
+buildoptions_vs_test = buildoptions_vs_app .. ' /wd4127 /D_WINSOCK_DEPRECATED_NO_WARNINGS'
```
Comment (sorgom) [inline make/premake5_vs.lua:22]: this must be commented

File: make/premake5_vs.lua [hunk 20]
```diff
@@ -32,29 +32,34 @@ buildoptions_vs_test = buildoptions_vs_app .. ' /wd4127'
 --  ============================================================
 workspace 'tests'
     filter { 'action:vs*' }
-        configurations { 'ci', 'dev' }
+        configurations { 'ci', 'dev', 'sys', 'qnd' }
         language 'C++'
         objdir 'obj/vs/%{prj.name}'
 
-        includedirs { includedirs_test }
+        includedirs { includedirs_test, includedirs_app }
```
Comment (sorgom) [inline make/premake5_vs.lua]: app should be already included

File: make/runSystemTests.sh [hunk 21]
```diff
@@ -0,0 +1,33 @@
+#!/bin/bash
+#   ====================================
+#   run real time tests
```
Comment (sorgom) [inline make/runSystemTests.sh]: system tests

File: specification/ifs/I_TCP.h [hunk 22]
```diff
@@ -0,0 +1,74 @@
...
+//  created by Manfred Sorgo
+
+#pragma once
+
+#include "DataTypes.h"
+
+class I_TCP
+{
+public:
+    //  set elect timeout in milli seconds
```
Comment (sorgom) [inline specification/ifs/I_TCP.h]: select

File: specification/ifs/values.h [hunk 23]
```diff
@@ -4,45 +4,40 @@
...
     COMP_LCR,
     COMP_SEG,
     COMP_SYS
 };
 
-//  ============================================================
-//  - return types
-//  ============================================================
+//  return types
 enum E_Ret : UINT8
```
Comment (sorgom) [inline specification/ifs/values.h:22]: these value should be decribed

File: specification/ifs/values.h [hunk 24]
```diff
@@ -98,14 +89,12 @@ enum E_TswState : UINT8
 
 enum E_TswGuiCmd : UINT8
```
Comment (sorgom) [inline specification/ifs/values.h]: E_TswCmd?

File: testing/testenv/TCP/TCP_Client.h [hunk 25]
```diff
@@ -0,0 +1,52 @@
...
+        void close();
+
+        inline bool result() const { return mRes; } 
+        //  set wait time in ms for thread based methods
+        //  should be less than TCP select timeout
+        inline void setWait(UINT32 ms) { mWait = ms; }
+        NOCOPY(TCP_Client)
+    private:
+        INT32 mSocket = -1;
+        std::atomic<bool> mRes = false;
```
Comment (sorgom) [inline testing/testenv/TCP/TCP_Client.h]: old crap

File: testing/testenv/TCP/TCP_Client.h [hunk 26]
```diff
@@ -0,0 +1,52 @@
...
+{
+    class TCP_Client
+    {
+    public:
+        TCP_Client() = default;
+        //  connect to localhost with port
+        //  works if test and app run in different processes
+        bool connect(UINT16 port);
+
+        bool send(CPTR data, INT32 size);
```
Comment (sorgom) [inline testing/testenv/TCP/TCP_Client.h]: comment on methods

File: testing/testenv/TCP/src/TCP_Client.cpp [hunk 27]
```diff
@@ -0,0 +1,82 @@
...
+#ifdef _WIN32
+#include <winsock2.h>
+#else
+#include <sys/socket.h>
+#include <arpa/inet.h>
+#include <unistd.h>
+#define closesocket close
+#endif
+
+// #include <thread>
```
Comment (sorgom) [inline testing/testenv/TCP/src/TCP_Client.cpp]: remove both lines

File: testing/testenv/TCP/src/TCP_Client.cpp [hunk 28]
```diff
@@ -0,0 +1,82 @@
...
+    void TCP_Client::close()
+    { 
+        if (mSocket >= 0) 
+        { 
+            ::closesocket(mSocket); 
+            mSocket = -1; 
+        } 
+    }
+
+    // void TCP_Client::t_connect(UINT16 port)
```
Comment (sorgom) [inline testing/testenv/TCP/src/TCP_Client.cpp]: remove crap

File: testing/testenv/mocks/M_Com.h [hunk 29]
```diff
@@ -18,6 +16,24 @@ namespace test
         inline M_Com() : M_Base("Com") {}
```
Comment (sorgom) [inline testing/testenv/mocks/M_Com.h]: avoid public constructor and instance

File: testing/testenv/mocks/M_Log.h [hunk 30]
```diff
@@ -13,7 +11,7 @@ namespace test
 {
     class M_Log : 
         public I_Log,
-        protected M_Base
+        private M_Base
     {
     public:
         M_Log() :
```
Comment (sorgom) [inline testing/testenv/mocks/M_Log.h]: instance / public con

File: testing/testenv/mocks/M_Reader.h [hunk 31]
```diff
@@ -0,0 +1,42 @@
...
+#include <ifs/I_Reader.h>
+#include "M_Base.h"
+#include <testlib/testValues.h>
+
+namespace test
+{
+    class M_Reader : public I_Reader, private M_Base
+    {
+    public:
+        inline M_Reader() : M_Base("Reader") {}
```
Comment (sorgom) [inline testing/testenv/mocks/M_Reader.h]: privatize!

File: testing/testenv/mocks/M_TCP.h [hunk 32]
```diff
@@ -0,0 +1,145 @@
...
+#include <ifs/I_TCP.h>
+#include "M_Base.h"
+#include <cstring>
+
+namespace test
+{
+    class M_TCP : public I_TCP, private M_Base
+    {
+    public:
+        inline M_TCP() : M_Base("Tcp") {}
```
Comment (sorgom) [inline testing/testenv/mocks/M_TCP.h]: privatize!

File: testing/testenv/mocks/M_TCP_Con.h [hunk 33]
```diff
@@ -0,0 +1,85 @@
...
+        NODEF(M_TCP_Con)
+    protected:
+        inline M_TCP_Con(const CONST_C_STRING name) : M_Base(name) {}
+
+    };
+
+    class M_TCP_Con_Fld : public M_TCP_Con
+    {
+    public:
+        inline M_TCP_Con_Fld() : M_TCP_Con("TCP_Con_Fld") {}
```
Comment (sorgom) [inline testing/testenv/mocks/M_TCP_Con.h]: privatize!

File: testing/testenv/mocks/M_TCP_Listener.h [hunk 34]
```diff
@@ -0,0 +1,76 @@
...
+    public:
+        inline M_TCP_Listener_Fld() : M_TCP_Listener("TCP_Listener_Fld") {}
+        INSTANCE_DEC(M_TCP_Listener_Fld)
+        NOCOPY(M_TCP_Listener_Fld)
+    };
+
+    class M_TCP_Listener_Gui : public M_TCP_Listener
+    {   
+    public:
+        inline M_TCP_Listener_Gui() : M_TCP_Listener("TCP_Listener_Gui") {}
```
Comment (sorgom) [inline testing/testenv/mocks/M_TCP_Listener.h]: privatize!

File: testing/testenv/mocks/ILPlugs.h
Comment (sorgom) [inline testing/testenv/mocks/ILPlugs.h:1]: - there is only 2 cases of unmock: SYST_01, SYST02
- these tests can be replaced by real system tests
- the whole plugging gets redundant
- IL.h: just forward to the mocks

File: testing/testenv/mocks/src/ILPlugs.cpp
Comment (sorgom) [inline testing/testenv/mocks/src/ILPlugs.cpp:1]: see comments on header

File: testing/testenv/mocks/M_Instances.h [hunk 35]
```diff
@@ -8,32 +8,51 @@
...
 #include "M_Provider.h"
+#include "M_Reader.h"
+#include "M_TCP_Con.h"
+#include "M_TCP_Listener.h"
+#include "M_TCP.h"
 
 namespace test
 {
-    inline M_Com& m_Com() { return M_Com::instance(); }
-    inline void mock_Com() { ILPLugs::instance().mCom.set(m_Com()); }
```
Comment (sorgom) [inline testing/testenv/mocks/M_Instances.h:24]: see comment on ILPlugs: not needed

File: testing/testenv/testlib/CppUTest.h [hunk 36]
```diff
@@ -7,8 +7,6 @@
 //  created by Manfred Sorgo
 
 #pragma once
-#ifndef CPPUTEST_H
-#define CPPUTEST_H
 
 //  no new malloc in application
 #define CPPUTEST_MEM_LEAK_DETECTION_DISABLED
```
Comment (sorgom) [inline testing/testenv/testlib/CppUTest.h:12]: do we need that? it's excluded by library build

File: testing/testenv/testlib/GenProjData.h [hunk 37]
```diff
@@ -39,6 +37,7 @@ namespace test
         ProjVec mTSWs, mSIGs, mLCRs, mSEGs;
 
     public:
+        const ComSetup mComSetup = { tcpPortFld, tcpPortGui, tcpPortCtrl, tcpTimeout };
```
Comment (sorgom) [inline testing/testenv/testlib/GenProjData.h:40]: maybe M_Reader and this can share

File: testing/testenv/testlib/TestLib.h [hunk 38]
```diff
@@ -4,30 +4,24 @@
...
+    //  preset ComName for testing
+    void genComName(ComName& cName, UINT32 num, CONST_C_STRING name = "ELM");
 
     template <class T>
     void nameElement(T& elem, UINT32 num, CONST_C_STRING name = "ELM")
     {
         genComName(elem.name, num, name);
     }
 
+    const ComName& genComName(UINT32 num, CONST_C_STRING name = "ELM");
```
Comment (sorgom) [inline testing/testenv/testlib/TestLib.h:23]: maybe extend this to all data types

File: testing/testenv/testlib/testValues.h [hunk 39]
```diff
@@ -31,4 +32,25 @@
 #define TEST_NUM_SEG 8
 #endif 
 
-#endif // H_
\ No newline at end of file
+namespace test
+{
+    // constexpr UINT16 tcpPortFld  = 8080;
```
Comment (sorgom) [inline testing/testenv/testlib/testValues.h]: remove crap

File: testing/tests/moduletests/COM/COM_01.cpp [hunk 40]
```diff
@@ -0,0 +1,162 @@
...
+    TEST(COM_01, T01)
+    {
+        I_TCP& tcp = TCP::instance();
+        tcp.setTimeout(5);
+        TCP_Client client;
+        INT32 res = 0;
+        bool ok = false;
+        CHAR buffer[sizeof(ComTele) + 10];
+
+        INT32 s1 = -1, s2 = -1;
```
Comment (sorgom) [inline testing/tests/moduletests/COM/COM_01.cpp]: some more speaking names?

File: testing/tests/moduletests/COM/COM_02.cpp [hunk 41]
```diff
@@ -0,0 +1,153 @@
...
+#include <testlib/TestGroupBase.h>
+#include <COM/TCP_Listeners.h>
+
+namespace test
+{
+
+    TEST_GROUP_BASE(COM_02, TestGroupBase)
+    {
+        constexpr static INT32 validSocket = 0;
+        constexpr static INT32 invalidSocket = -1;
```
Comment (sorgom) [inline testing/tests/moduletests/COM/COM_02.cpp:15]: is this used?

File: testing/tests/moduletests/COM/COM_03.cpp [hunk 42]
```diff
@@ -0,0 +1,228 @@
...
+        CHECK_N_CLEAR()
+        L_CHECK_TRUE(ok)
+
+        //  select returns 1 activity
+        //  recv returns valid data
+        //  Dispatcher gets data from GUI
+        STEP(2)
+        {
+            const ComTele tele{{}, {111, 212}};
+            m_TCP().expectSelect(validSocket, 1);
```
Comment (sorgom) [inline testing/tests/moduletests/COM/COM_03.cpp:115]: in general there should be an enum for select returns

File: testing/tests/moduletests/SYS/SYS_01.cpp [hunk 43]
```diff
@@ -114,4 +115,11 @@ namespace test
         I_Reader& inst = Reader::instance();
         play(inst);
     }
+
+    //  test type: coverage
+    //  log instance
+    TEST(SYS_01, T06)
+    {
+        Log::instance().log(COMP_SYS, RET_ERR_MATCH);
```
Comment (sorgom) [inline testing/tests/moduletests/SYS/SYS_01.cpp]: do not forget to remove SYS_03.cpp

File: testing/tests/moduletests/SYS/SYS_02.cpp [hunk 44]
```diff
@@ -1,5 +1,5 @@
 //  ============================================================
-//  test of module Dispatcher
+//  test of modules Dispatcher
```
Comment (sorgom) [inline testing/tests/moduletests/SYS/SYS_02.cpp]: test of Dispatcher?

File: testing/tests/moduletests/SYS/SYS_03.cpp
Comment (sorgom) [inline testing/tests/moduletests/SYS/SYS_03.cpp:1]: remove completely

File: testing/tests/runtests/RUNT_01.cpp
Comment (sorgom) [inline testing/tests/runtests/RUNT_01.cpp:1]: remove old systests, then rename to SYST_01

File: testing/tests/runtests/RUNT_01.cpp [hunk 45]
```diff
@@ -0,0 +1,135 @@
+//  ============================================================
+//  runtime tests vol. 1
```
Comment (sorgom) [inline testing/tests/runtests/RUNT_01.cpp]: system tests

File: testing/tests/runtests/RUNT_01.cpp [hunk 46]
```diff
@@ -0,0 +1,135 @@
...
+        L_CHECK_TRUE(ok)
+
+        //  receive switch state telegram to field
+        //  receive wait state telegram to GUI
+        STEP(2)
+        const ComTele txf{cname, { TSW_STATE_RIGHT, PARAM_UNDEF }};
+        const ComTele txg{cname, { TSW_STATE_WAIT_RIGHT, PARAM_UNDEF }};
+        ComTele tr;
+        
+        ok = clientFld.recv(tr);
```
Comment (sorgom) [inline testing/tests/runtests/RUNT_01.cpp]: reduce code by methods

File: testing/tests/systemtests/SYST_01.cpp
Comment (sorgom) [inline testing/tests/systemtests/SYST_01.cpp:1]: burn after reading

File: testing/tests/systemtests/SYST_02.cpp
Comment (sorgom) [inline testing/tests/systemtests/SYST_02.cpp:1]: burn after reading

File: make/premake5.lua (hunk 14 above)
Comment (sorgom) [reply make/premake5.lua:34]: made it an issue

File: make/premake5_settings.lua (hunk 17 above)
Comment (sorgom) [reply make/premake5_settings.lua:67]: left it the same, applied on system tests

File: make/premake5_vs.lua (hunk 20 above)
Comment (sorgom) [reply make/premake5_vs.lua]: correct

File: make/premake5_vs.lua (hunk 19 above)
Comment (sorgom) [reply make/premake5_vs.lua:22]: postponed

File: testing/testenv/testlib/CppUTest.h (hunk 36 above)
Comment (sorgom) [reply testing/testenv/testlib/CppUTest.h:12]: postponed to issue  #60

File: testing/testenv/testlib/GenProjData.h (hunk 37 above)
Comment (sorgom) [reply testing/testenv/testlib/GenProjData.h:40]: would require an extern and a source file; too complicated

File: testing/testenv/testlib/TestLib.h (hunk 38 above)
Comment (sorgom) [reply testing/testenv/testlib/TestLib.h:23]: -> ComData, ComTele

File: testing/tests/moduletests/COM/COM_01.cpp (hunk 40 above)
Comment (sorgom) [reply testing/tests/moduletests/COM/COM_01.cpp]: now: socket1, socket2

File: testing/tests/moduletests/COM/COM_02.cpp (hunk 41 above)
Comment (sorgom) [reply testing/tests/moduletests/COM/COM_02.cpp:15]: yes

File: testing/tests/moduletests/COM/COM_03.cpp (hunk 42 above)
Comment (sorgom) [reply testing/tests/moduletests/COM/COM_03.cpp:115]: postponed to issue #61

File: testing/tests/runtests/RUNT_01.cpp (hunk 46 above)
Comment (sorgom) [reply testing/tests/runtests/RUNT_01.cpp]: introduced M_TCP_Client

File: make/README.md
Comment (sombeam) [inline make/README.md:1]: comment on runSystemTests also for gcc

File: make/premake5.lua (hunk 14 above)
Comment (sombeam) [inline make/premake5.lua:34]: this must be at least commented

File: make/premake5.lua [hunk 47]
```diff
@@ -29,7 +29,7 @@ workspace 'lib_cpputest'
 --  configurations: 
 --  - ci        module tests
```
Comment (sombeam) [inline make/premake5.lua:41]: reduce to ci and debug

File: make/premake5.lua [hunk 48]
```diff
@@ -82,7 +84,7 @@ workspace 'tests'
 --  ============================================================
 workspace 'coverage'
     filter { 'action:gmake*' }
```
Comment (sombeam) [inline make/premake5.lua:101]: ci should be enough

File: make/premake5.lua [hunk 49]
```diff
@@ -97,60 +99,57 @@ workspace 'coverage'
...
---  -   system tests only runtime
---  ->  bin/sysests_tests_{config}
+--  > systemtests.make
+--  -   run tests only runtime
 --  configurations: 
 --  - ci        module tests
 --  - qnd       with devel includes
 --  ============================================================
-workspace 'systests'
+workspace 'systemtests'
```
Comment (sombeam) [inline make/premake5.lua:132]: remove the qnd stuff?

File: make/premake5.lua
Comment (sombeam) [inline make/premake5.lua:1]: we should generally have ci and debug

File: make/premake5.lua [hunk 50]
```diff
@@ -168,12 +167,30 @@ workspace 'dstw'
...
 
         project 'dstw_run'
             kind 'ConsoleApp'
             includedirs { includedirs_app }
             defines { defines_app }
-            files { files_app }
+            files { files_app, files_app_main }
+
+--  ============================================================
+--  > _valgrind.make
```
Comment (sombeam) [inline make/premake5.lua]: mention valgrind in markdown

File: make/premake5_settings.lua [hunk 51]
```diff
@@ -58,16 +65,20 @@ defines_test = {
 defines_app = { 'NDEBUG' }
 
 defines_gendata = { 
-    'NDEBUG', 
-    'TEST_NUM_TSW=1000', 
-    'TEST_NUM_SIG=1000', 
-    'TEST_NUM_LCR=1000', 
-    'TEST_NUM_SEG=1000' 
+    defines_test,
```
Comment (sombeam) [inline make/premake5_settings.lua:68]: use defaults, remove here

File: make/premake5_vs.lua (hunk 19 above)
Comment (sombeam) [inline make/premake5_vs.lua:22]: comment!

File: make/premake5_vs.lua [hunk 52]
```diff
@@ -32,7 +32,7 @@ buildoptions_vs_test = buildoptions_vs_app .. ' /wd4127'
 --  ============================================================
 workspace 'tests'
     filter { 'action:vs*' }
-        configurations { 'ci', 'dev' }
+        configurations { 'ci', 'dev', 'qnd' }
```
Comment (sombeam) [inline make/premake5_vs.lua]: reduce to ci and debug

File: make/premake5_vs.lua [hunk 53]
```diff
@@ -44,58 +44,64 @@ workspace 'tests'
...
+--  build all executables at once:
+--  - application runtime
+--  - generate proj data for runtime
+--  - TCP based system tests
+--  - TCP based application stop
 --  ============================================================
-workspace 'gendata'
+workspace 'dstw_system'
     filter { 'action:vs*' }
         configurations { 'ci' }
```
Comment (sombeam) [inline make/premake5_vs.lua]: ci and debug

File: make/runGcov.sh [hunk 54]
```diff
@@ -4,6 +4,14 @@
 #   ====================================
 cd $(dirname $0)
 
+show=0
+while getopts s option; do
+    case $option in
```
Comment (sombeam) [inline make/runGcov.sh:20]: help?

File: specification/ifs/values.h [hunk 55]
```diff
@@ -4,45 +4,46 @@
...
-//  - parameter values
-//  ============================================================
+//  reserved parameter values
 enum E_Param : UINT8
 {
-    PARAM_UNDEF = 255
+    //  value not defined
+    PARAM_UNDEF = 254,
+    //  value recall to field
+    PARAM_RECALL = 255
```
Comment (sombeam) [inline specification/ifs/values.h]: unused

File: testing/README.md [hunk 56]
```diff
@@ -1,37 +1,100 @@
...
 
-In system test context only exit points of the system are mocked.
-Triggers take place at the system entry points. 
+In test context every module instance that realizes an interface is substituted by a mock.
 
 The software compiled for production does not provide any substitution of the modules in charge.
+
+The whole thing is achieved by different include paths leading to two different IL.h headers.
+
+#### header for production build
```
Comment (sombeam) [inline testing/README.md]: just link to headers, maybe at end

File: testing/tests/moduletests/SYS/SYS_01.cpp [hunk 57]
```diff
@@ -103,15 +111,14 @@ namespace test
...
         STEP(1)
         expectFail();
-        mSUT.read(fname);
+        rdr.read(fname);
         CHECK_N_CLEAR()
     }
 
     //  test type: coverage
-    //  retrieve instance
+    //  log instance
```
Comment (sombeam) [inline testing/tests/moduletests/SYS/SYS_01.cpp]: why don't we test maxerr?

**Review by sombeam — CHANGES_REQUESTED** [review]

File: make/premake5.lua (hunk 14 above)
Comment (sorgom) [reply make/premake5.lua:34]: postponed due to don't know suitable comment

File: application/components/SYS/src/Reader.cpp (hunk 13 above)
Comment (sorgom) [reply application/components/SYS/src/Reader.cpp:26]: postponed

File: make/premake5.lua (hunk 50 above)
Comment (sorgom) [reply make/premake5.lua]: removed valgrind stuff

File: make/premake5_vs.lua (hunk 19 above)
Comment (sorgom) [reply make/premake5_vs.lua:22]: postponed

File: testing/tests/moduletests/SYS/SYS_01.cpp [hunk 58]
```diff
@@ -1,16 +1,16 @@
 //  ============================================================
-//  test of module Reader
+//  test of module Reader, Log (coverage)
```
Comment (sombeam) [inline testing/tests/moduletests/SYS/SYS_01.cpp]: remove: (coverage)

**Review by sombeam — CHANGES_REQUESTED** [review]

**Review by sombeam — APPROVED** [review]

**Pull request merged**